pkg/devicemapper/devmapper.go

```diff
@@ -361,15 +361,13 @@ func RemoveDeviceDeferred(name string) error {
 	// set a task cookie and disable library fallback, or else libdevmapper will
 	// disable udev dm rules and delete the symlink under /dev/mapper by itself,
 	// no matter the removal is executed immediately or deferred by the kernel.
-	// Calling UdevWait on the cookie after task.run is not necessary, since deferred
-	// device removal is an asynchronous api call.
 	var cookie uint
 	var flags uint16
 	flags = DmUdevDisableLibraryFallback
-
 	if err := task.setCookie(&cookie, flags); err != nil {
 		return fmt.Errorf("devicemapper: Can not set cookie: %s", err)
 	}
+	defer UdevWait(&cookie)
 
```

Comment:
libdevmapper use system-v semaphore for synchronization with udev, and the cookie(which is a random string generated when dm_task_set_cookie get called) is used as the semaphore's id.Before libdevmapper do the `ioctl` syscall, it first create a semaphore with the cookie as id, and increment it by one, after the syscall, it checks `ioctl`'s return value to see whether a `DM_UEVENT_GENERATED_FLAG`, if not set, libdevmapper knows there is no need to wait for udev completing its job, so it will decrement the semaphore immediately, if set, libdevmapper knows there is a udev event generated by kernel, and udev have some work to do for that event.When udev catches that event, it (I guess) will increment the semaphore first and do its jobs against udev rules, and finally decrement the semaphore.What libdevmapper actually do in `dm_udev_wait` is undo its semaphore increment first, and repeatedly wait for the semaphore's value decrements to zero.

For our case, if the target device is not busy, the kernel will remove it immediately, which acts the same as a normal remove operation, so call `UdevWait` totally make sense. When the device is busy, kernel will mark the device as deferred remove, and no udev event generated, the semaphore's value will be decremented right after the `ioctl` syscall, and what `UdevWait` do later is merely nothing but cleaning up the semaphore.